src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,88 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    //
+    // For Amd64 System V, a two-register-returned struct could be returned in RAX and RDX
+    // In such case make sure that the correct GC-ness of RDX is reported as well, so
+    // a GC object pointed by RDX will not be collected.
+    if (!pushReg)
+    {
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        // Handling struct returned in two registers (only applicable to System V systems)...
+        if (compiler->compMethodReturnsMultiRegRetType())
+        {
+            // Get the return tye of the struct.
+            SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR retStructDesc;
+            compiler->eeGetSystemVAmd64PassStructInRegisterDescriptor(compiler->info.compMethodInfo->args.retTypeClass, &retStructDesc);
+
+            assert(retStructDesc.passedInRegisters);
+
+            // In case the return type is a two-register-return, the native return type should be a struct
+            assert(varTypeIsStruct(compiler->info.compRetNativeType));
+
+            assert(retStructDesc.eightByteCount == CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
+
+            unsigned __int8 offset0 = 0;
+            unsigned __int8 offset1 = 0;
+
+            var_types type0 = TYP_UNKNOWN;
+            var_types type1 = TYP_UNKNOWN;
+
+            // Set the GC-ness of the struct return registers.
+            getStructTypeOffset(retStructDesc, &type0, &type1, &offset0, &offset1);
+            gcInfo.gcMarkRegPtrVal(REG_INTRET, type0);
+            gcInfo.gcMarkRegPtrVal(REG_INTRET_1, type1);
+        }
+        // ... handling return in implicit RetBuf and returning the RetBuf address in REG_INTRET...
+        else
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (compiler->compMethodReturnsRetBufAddr())
+        {
+            // This is for returning in an implicit RetBuf.
+            // If the address of the buffer is returned in REG_INTRET, mark the content of INTRET as ByRef.
+
+            // In case the return is in an implicit RetBuf, the native return type should be a struct
+            assert(varTypeIsStruct(compiler->info.compRetNativeType));
+
+            gcInfo.gcMarkRegPtrVal(REG_INTRET, TYP_BYREF);
+        }
+        // ... all other cases.
+        else
+        {
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+            // For System V the above if statement (if (compiler->compMethodReturnsRetBufAddr()))
+            // should always evaluate to true since the ABI requires to return the 
+            // address of the RetBuf in REG_INTRET.
+            assert(compiler->info.compRetBuffArg == BAD_VAR_NUM);
```

Comment:
The comment and the assert do not match. If the above statement always evaluates to true, then we should NEVER get here. I think that what the comment should be saying is that for structs that are not returned in registers, they should always be returned in the return buffer. Therefore, if we reach here, we should not have a return buffer, and the return type should not be struct.